clang-tools-extra/clang-tidy/readability/UseStdMinMaxCheck.cpp

```diff
@@ -0,0 +1,137 @@
+//===--- UseStdMinMaxCheck.cpp - clang-tidy -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "UseStdMinMaxCheck.h"
+#include "../utils/ASTUtils.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/Lex/Preprocessor.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::readability {
+
+static const llvm::StringRef AlgorithmHeader("<algorithm>");
+
+UseStdMinMaxCheck::UseStdMinMaxCheck(StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context),
+      IncludeInserter(Options.getLocalOrGlobal("IncludeStyle",
+                                               utils::IncludeSorter::IS_LLVM),
+                      areDiagsSelfContained()) {}
+
+void UseStdMinMaxCheck::storeOptions(ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, "IncludeStyle", IncludeInserter.getStyle());
+}
+
+void UseStdMinMaxCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      ifStmt(
+          stmt().bind("if"),
+          hasCondition(binaryOperator(hasAnyOperatorName("<", ">", "<=", ">="),
+                                      hasLHS(expr().bind("CondLhs")),
+                                      hasRHS(expr().bind("CondRhs")))),
+          hasThen(anyOf(stmt(binaryOperator(hasOperatorName("="),
+                                            hasLHS(expr().bind("AssignLhs")),
+                                            hasRHS(expr().bind("AssignRhs")))),
+                        compoundStmt(statementCountIs(1),
+                                     has(binaryOperator(
+                                         hasOperatorName("="),
+                                         hasLHS(expr().bind("AssignLhs")),
+                                         hasRHS(expr().bind("AssignRhs"))))))),
+          hasParent(stmt(unless(ifStmt(hasElse(
```

Comment:
This will prevent detecting code like this:
```
if (x)
  doSomething()
else
   if (a > b)
      a = b;
```

Where it could be valid to change this into:
```
if (x)
  doSomething()
else
   a = std::min(a, b);
```